File: 4-Sliding Window/3-character_replacement.cpp
#include <iostream>
#include <string>
#include <unordered_map>

using namespace std;

int characterReplacement(string s, int k) {
    int res = 0;
    int len = s.size();
    unordered_map<char, int> count;
    int left = 0;
    int maxFreq = 0;
    
    for (int right = 0; right < len; right++) {
        count[s[right]]++;
        maxFreq = max(maxFreq, count[s[right]]);

        while ((right - left + 1) - maxFreq > k) {
            count[s[left]]--;
            left++;
        }
        res = max(res, right - left + 1);
    }
    return res;
}

int main() {
    string s = "AABABBA";
    
    cout << characterReplacement(s, 2);

    return 0;
}

File: PYTHON/4-SlidingWindow/3-character_replacement.py

def characterReplacement(s: str, k: int) -> int:
    res = 0
    char_set = set(s)

    for c in char_set:
        cont = left = 0
        for right in range(len(s)):
            if s[right] == c:
                cont += 1
            
            while (right - left + 1) - cont > k:
                if s[left] == c:
                    cont -= 1
                
                left += 1
            
            res = max(res, right - left + 1)
        
    return res

if __name__ == "__main__":
    print(characterReplacement(s="XYYX", k=2))